dtm0/drlink.c

```diff
@@ -466,27 +522,73 @@ static void drlink_coro_fom_tick(struct m0_co_context *context)
 					      DRF_LOCKING));
 	M0_ASSERT(m0_long_is_write_locked(&F(proc)->dop_llock, fom));
 
-	if (dpr_state_infer(F(proc)) == DPR_TRANSIENT) {
+	m0_conf_cache_lock(&confc->cc_cache);
+	obj = m0_conf_cache_lookup(&confc->cc_cache, &F(proc)->dop_rserv_fid);
+	M0_ASSERT(obj != NULL);
+	if (has_volatile_param(obj) && obj->co_ha_state != M0_NC_ONLINE) {
+		M0_LOG(M0_DEBUG, "Force state transition %s -> %s for " FID_F,
+		       m0_ha_state2str(obj->co_ha_state),
+		       m0_ha_state2str(obj->co_ha_state),
+		       FID_P(&F(proc)->dop_rserv_fid));
+		obj->co_ha_state = M0_NC_ONLINE;
+		m0_chan_broadcast(&obj->co_ha_chan);
+		always_reconnect = true;
+	}
+	m0_conf_cache_unlock(&confc->cc_cache);
+
+	/* Reconnect if the session was canceled */
+	if (m0_rpc_link_is_connected(&F(proc)->dop_rlink) &&
+	    F(proc)->dop_rlink.rlk_sess.s_cancelled) {
+		always_reconnect = true;
+	}
+
+	/* XXX:
+	 * At this moment we cannot detect client falures.
+	 * Because of that, we cannot detect the case where
+	 * the client drops RPC items because it cannot
+	 * find the corresponding connection.
+	 * As a workaround, we force drlink to re-connect
+	 * whenever it tries to send a message.
+	 */
+
+	if (always_reconnect) {
```

Comment:
So, in two cases we will do always_reconnect:
1. the client is marked as not online.
2. the rpc link is connected but is cancelled.

In these cases, we do reconnect before posting recovery messages.
Is that right?